src/ReservoirPriceOracle.sol

```diff
@@ -1,32 +1,405 @@
+        return Buffer.SIZE;
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+    //                                 INTERNAL FUNCTIONS                                        //
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+
+    function _validatePair(ReservoirPair aPair) internal pure {
+        if (address(aPair) == address(0)) revert NoDesignatedPair();
+    }
+
+    // TODO: replace this with safe, audited lib function
+    function _calcPercentageDiff(uint256 aOriginal, uint256 aNew) internal pure returns (uint256) {
+        if (aOriginal == 0) return 0;
+
+        if (aOriginal > aNew) {
+            return (aOriginal - aNew) * 1e18 / aOriginal;
+        } else {
+            return (aNew - aOriginal) * 1e18 / aOriginal;
+        }
+    }
+
+    function _rewardUpdater(address lRecipient) internal {
+        // N.B. Revisit this whenever deployment on a new chain is needed
+        // we use `block.basefee` instead of `ArbGasInfo::getMinimumGasPrice()` on ARB because the latter will always return
+        // the demand insensitive base fee, while the former can return real higher fees during times of congestion
+        // safety: this mul will not overflow even in extreme cases of `block.basefee`
+        uint256 lPayoutAmt = block.basefee * rewardMultiplier;
+
+        if (lPayoutAmt <= address(this).balance) {
+            payable(lRecipient).transfer(lPayoutAmt);
+        } else { } // do nothing if lPayoutAmt is greater than the balance
+    }
+
+    function _getQuote(uint256 aAmount, address aBase, address aQuote) internal view returns (uint256 rOut) {
+        (address lToken0, address lToken1) = aBase.sortTokens(aQuote);
+
+        address[] memory lRoute = _route[lToken0][lToken1];
+        if (lRoute.length == 0) revert PO_NoPath();
+
+        uint256 lPrice = WAD;
+        for (uint256 i = 0; i < lRoute.length - 1; ++i) {
+            // we need to sort token addresses again since intermediate path addresses are not guaranteed to be sorted
+            (address lLowerToken, address lHigherToken) = lRoute[i].sortTokens(lRoute[i + 1]);
+
+            // it is assumed that subroutes defined here are simple routes and not composite routes
+            // meaning, each segment of the route represents a real price between pair, and not the result of composite routing
+            // therefore we do not check `_route` again to ensure that there is indeed a route
+            uint256 lRoutePrice = priceCache[lLowerToken][lHigherToken];
+            lPrice = lPrice * (lLowerToken == lRoute[i] ? lRoutePrice : lRoutePrice.invertWad()) / WAD;
+        }
+
+        // idea: can build a cache of decimals to save on making external calls?
+        uint256 lBaseDecimals = IERC20(aBase).decimals();
+        uint256 lQuoteDecimals = IERC20(aQuote).decimals();
```

Comment:
Can we not just bake the decimals into the price we store? I.e. scale the price by 10**(18 -decimals)?

Reply:
the price returned by our oracle is scaled by `WAD`. During the price calculations in the AMM all token balances are normalized to `WAD` as well. Esp for the stable pair it's not really feasible to calculate prices without doing this scaling

Another reason this won't work is, when we multiply prices from different paths, if they're scaled differently this multiplication won't work. Right now it works because everything is scaled by 1e18. Your suggestion seems to only work when it's all simple direct A->B routes